File: LeetCodePractice/src/Contest/GetEqualSubstringsWithinBudget.java
package Contest;
/*You are given two strings s and t of the same length. You want to change s to t. Changing the i-th character of s to i-th character of t costs |s[i] - t[i]| that is, the absolute difference between the ASCII values of the characters.
You are also given an integer maxCost.
Return the maximum length of a substring of s that can be changed to be the same as the corresponding substring of twith a cost less than or equal to maxCost.
If there is no substring from s that can be changed to its corresponding substring from t, return 0.

Example 1:
Input: s = "abcd", t = "bcdf", maxCost = 3
Output: 3
Explanation: "abc" of s can change to "bcd". That costs 3, so the maximum length is 3.
Example 2:

Input: s = "abcd", t = "cdef", maxCost = 3
Output: 1
Explanation: Each character in s costs 2 to change to charactor in t, so the maximum length is 1.
Example 3:

Input: s = "abcd", t = "acde", maxCost = 0
Output: 1
Explanation: You can't make any change, so the maximum length is 1.
 
Constraints:
1 <= s.length, t.length <= 10^5
0 <= maxCost <= 10^6
s and t only contain lower case English letters.*/
public class GetEqualSubstringsWithinBudget {
	public int equalSubstring(String s, String t, int maxCost) {
		int l = 0, res = 0, cost = 0;
        for(int r = 0; r < s.length(); r++) {
        	cost += Math.abs(s.charAt(r) - t.charAt(r));
        	while(cost > maxCost) {
        		cost -= Math.abs(s.charAt(l) - t.charAt(l));
                l++;
        	}
        	res = Math.max(res, r - l + 1);
        }
        return res;
    }
}

File: LeetCodePython/GetEqualSubstringsWithinBudget.py
"""
You are given two strings s and t of the same length and an integer maxCost.

You want to change s to t. Changing the ith character of s to ith character of t costs |s[i] - t[i]| (i.e., the absolute difference between the ASCII values of the characters).

Return the maximum length of a substring of s that can be changed to be the same as the corresponding substring of t with a cost less than or equal to maxCost. If there is no substring from s that can be changed to its corresponding substring from t, return 0.

Example 1:
Input: s = "abcd", t = "bcdf", maxCost = 3
Output: 3
Explanation: "abc" of s can change to "bcd".
That costs 3, so the maximum length is 3.

Example 2:
Input: s = "abcd", t = "cdef", maxCost = 3
Output: 1
Explanation: Each character in s costs 2 to change to character in t,  so the maximum length is 1.

Example 3:
Input: s = "abcd", t = "acde", maxCost = 0
Output: 1
Explanation: You cannot make any change, so the maximum length is 1.

Constraints:
1 <= s.length <= 10^5
t.length == s.length
0 <= maxCost <= 10^6
s and t consist of only lowercase English letters.

hints:
1 Calculate the differences between a[i] and b[i].
2 Use a sliding window to track the longest valid substring.
"""

class Solution:
    def equalSubstring(self, s: str, t: str, maxCost: int) -> int:
        sz = len(s)
        l = cost = res = 0
        for r in range(sz):
            cost += abs(ord(s[r]) - ord(t[r]))
            if cost > maxCost:
                cost -= abs(ord(s[l]) - ord(t[l]))
                l += 1
            res = max(res, r - l + 1)
        return res